pkg/operator/target_status.go

```diff
@@ -0,0 +1,936 @@
+type targetStatusReconciler struct {
+	ch            chan event.GenericEvent
+	opts          Options
+	manager       dataManager
+	tickerBuilder func() Ticker
+	logger        logr.Logger
+}
+
+// newTargetStatusReconciler creates a new targetStatusReconciler and kicks off
+// the reconcile loop by sending the first event to the source channel.
+func NewTargetStatusPoller(logger logr.Logger, restConfig *rest.Config, manager manager.Manager, registry prometheus.Registerer, opts Options) (TargetStatusProcessor, error) {
+	dataManagerBuilder, err := newControllerDataManagerBuilder(restConfig, opts, manager.GetClient())
+	if err != nil {
+		return nil, err
+	}
+
+	reconciler, err := newTargetStatusPoller(logger, opts, registry, dataManagerBuilder)
+	err = ctrl.NewControllerManagedBy(manager).
+		Named("target-status").
+		For(
+			// We don't actually listen to any DaemonSet, just the channel.
+			&appsv1.DaemonSet{},
+			builder.WithPredicates(predicate.NewPredicateFuncs(func(o client.Object) bool {
+				return false
+			})),
+		).
+		Watches(&source.Channel{
+			Source: reconciler.ch,
+		}, &handler.EnqueueRequestForObject{}).
+		Complete(reconciler)
+	if err != nil {
+		return nil, errors.Wrap(err, "create target status controller")
+	}
+
+	return reconciler, nil
+}
+
+func newTargetStatusPoller(logger logr.Logger, opts Options, registry prometheus.Registerer, builder dataManagerBuilder) (*targetStatusReconciler, error) {
+	if err := registry.Register(targetStatusDuration); err != nil {
+		return nil, err
+	}
+
+	ch := make(chan event.GenericEvent, 1)
+
+	dataManager, err := builder.build()
+	if err != nil {
+		return nil, err
+	}
+
+	reconciler := &targetStatusReconciler{
+		ch:      ch,
+		opts:    opts,
+		manager: dataManager,
+		logger:  logger,
+	}
+
+	return reconciler, nil
+}
+
+type emptyObject struct{}
+
+func (r *targetStatusReconciler) Start(ctx context.Context, tickerBuilder func() Ticker) error {
+	r.tickerBuilder = tickerBuilder
+	r.queue()
+	return nil
+}
+
+// Queues a reconcile.
+func (r *targetStatusReconciler) queue() {
+	r.ch <- event.GenericEvent{
+		Object: &appsv1.DaemonSet{},
+	}
+}
+
+// Reconcile polls the collector pods, fetches and aggregates target status and upserts
+// into each PodMonitoring's Status field.
+func (r *targetStatusReconciler) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	// minDelay := time.NewTimer(10 * time.Second)
+
+	minDelay := r.tickerBuilder()
+	// defer minDelay.Stop()
+
+	onceLog := new(sync.Once)
+
+	// Wrapper function so we could exit early.
+	err := func() error {
+		now := time.Now()
+		if err := poll(ctx, r.logger, r.opts, r.manager); err != nil {
+			// Not having data is not an error.
+			if apierrors.IsNotFound(err) {
+				onceLog.Do(func() {
+					r.logger.Error(err, "Kubernetes config not found")
+				})
+				return nil
+			}
+			return err
+		}
+		// In case Kubernetes config is removed and re-added.
```

Comment:
Realistically rarely. It's primarily helpful to self-deployed collection users or developers who fiddle with configurations, e.g. the collector DaemonSet. The `Once` helps prevent flooding the logs with error messages every tick, but also resets in case the errors reoccur later on. This error might also happen if somehow the Kubernetes controller watcher goes out of sync.